### PR TITLE
Error checking for non-iterables

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes https://github.com/Lightning-AI/lightning/issues/16912

cc @borda @justusschock @awaelchli

### CODE REVIEW

File: src/lightning/pytorch/trainer/connectors/data_connector.py [hunk 1]
```diff
@@ -264,15 +251,18 @@ def _get_distributed_sampler(
 
 
 def _resolve_overfit_batches(combined_loader: CombinedLoader, mode: RunningStage) -> None:
-    all_have_sequential_sampler = all(isinstance(dl.sampler, SequentialSampler) for dl in combined_loader.flattened)
+    all_have_sequential_sampler = all(
+        isinstance(dl.sampler, SequentialSampler) for dl in combined_loader.flattened if hasattr(dl, "sampler")
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/connectors/data_connector.py:255]: This is a small bugfix for using non-dataloaders with overfit_batches enabeld

File: src/lightning/pytorch/trainer/connectors/data_connector.py [hunk 2]
```diff
@@ -303,11 +293,9 @@ def dataloader(self) -> Union[TRAIN_DATALOADERS, EVAL_DATALOADERS]:
         """
         if isinstance(self.instance, pl.LightningModule):
             return call._call_lightning_module_hook(self.instance.trainer, self.name, pl_module=self.instance)
-
         if isinstance(self.instance, pl.LightningDataModule):
-            method = getattr(self.instance, self.name)
-            return method()
-
+            assert self.instance.trainer is not None
+            return call._call_lightning_datamodule_hook(self.instance.trainer, self.name)
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/connectors/data_connector.py:298]: I don't know why we weren't using this before. Maybe missed when it was added

File: tests/tests_pytorch/loops/test_loops.py [hunk 3]
```diff
@@ -786,29 +806,58 @@ def _get_iterator(self):
...
-    assert train_dataloader.count_shutdown_workers == 2 if should_fail else (2 if persistent_workers else max_epochs)
-    # on sanity checking end, the workers are being deleted too.
-    expected = 2 if persistent_workers else (3 if should_fail else max_epochs + 1)
-    assert val_dataloader.count_shutdown_workers == expected
+    if persistent_workers:
+        expected = [trainer.current_epoch, trainer.current_epoch]  # once epoch end, once on teardown
+    elif should_fail:
+        expected = [
+            # iterable check
+            0,
```
Comment (carmocca) [inline tests/tests_pytorch/loops/test_loops.py:820]: This extra shutdown call is caused by the added iterable checks. This is the only change from master in this test

This means that checking whether the input dataloader is iterable does have a performance impact.

**Review by Borda — APPROVED** [review]

File: src/lightning/pytorch/trainer/connectors/data_connector.py [hunk 4]
```diff
@@ -386,18 +374,30 @@ def get_instance(self, hook_name: str) -> Union["pl.LightningModule", "pl.Lightn
...
+                    f" Either pass the dataloader to the `.{trainer_fn}()` method OR implement"
+                    f" `def {source.name}(self):` in your LightningModule/LightningDataModule."
+                )
+            raise TypeError(
+                f"An invalid dataloader was returned from `{type(source.instance).__name__}.{source.name}()`."
+                f" Found {dataloader}."
+            )
+        raise TypeError(
+            f"An invalid dataloader was passed to `Trainer.{trainer_fn}({prefix}dataloaders=...)`. Found {dataloader}."
         )
```
Comment (awaelchli) [inline src/lightning/pytorch/trainer/connectors/data_connector.py:401]: Nit: You could avoid the extra indentation by doing

```py
if not source.is_module():
    raise ...
if not is_overridden()
    raise ...
raise ...
```

File: tests/tests_pytorch/trainer/connectors/test_data_connector.py [hunk 5]
```diff
@@ -614,8 +614,34 @@ def test_attach_data_input_validation_with_none_dataloader(trainer_fn_name, data
...
+@pytest.mark.parametrize(
+    "trainer_fn_name, dataloader_name, stage",
+    [
+        ("fit", "train_dataloaders", RunningStage.TRAINING),
+        ("validate", "dataloaders", RunningStage.VALIDATING),
+        ("test", "dataloaders", RunningStage.TESTING),
+        ("predict", "dataloaders", RunningStage.PREDICTING),
+    ],
+)
+@pytest.mark.parametrize("dataloader", [object(), [1, object()]])
```
Comment (awaelchli) [inline tests/tests_pytorch/trainer/connectors/test_data_connector.py:633]:
```suggestion
@pytest.mark.parametrize("dataloader", [None, object(), [1, object()]])
```

**Review by awaelchli — APPROVED** [review]

**Pull request merged**